File: python/problem_33.py
# Problem 33 (Digit cancelling fractions)
# =======================================
#
# The fraction 49/98 is a curious fraction, as an inexperienced mathematician
# in attempting to simplify it may incorrectly believe that 49/98 = 4/8, which
# is correct, is obtained by cancelling the 9s.
#
# We shall consider fractions like, 30/50 = 3/5, to be trivial examples.
#
# There are exactly four non-trivial examples of this type of fraction, less
# than one in value, and containing two digits in the numerator and denominator.
#
# If the product of these four fractions is given in its lowest common terms,
# find the value of the denominator.

# We're looking for fractions:
#  - without both numerator & denominator ending in same number (non-trivial)
#  - less than one (numerator < denominator)
#  - two digits in numerator and denominator (max. (99-9)² possibilities)
from fractions import Fraction
from common import product


if __name__ == "__main__":
    fractions = []

    for numerator in range(10,100):
        a = numerator // 10
        b = numerator % 10

        for denominator in range(10,100):
            c = denominator // 10
            d = denominator % 10

            if numerator >= denominator:  # must be < 1
                continue

            if b == d:  # must be non-trivial
                continue

            # curiously simplified fraction
            try:
                if a == d:
                    curiously_simplified = Fraction(b, c)
                elif b == c:
                    curiously_simplified = Fraction(a, d)
                elif a == c:
                    curiously_simplified = Fraction(b, d)
                else:
                    continue  # definitely not a curious fraction
            except ZeroDivisionError:
                continue

            # properly simplified fraction
            f = Fraction(numerator, denominator)

            if f == curiously_simplified:
                fractions.append(f)

    print(product(fractions).denominator)

File: haskell/problem_33.hs
-- Problem 33 (Digit cancelling fractions)
-- =======================================
--
-- The fraction 49/98 is a curious fraction, as an inexperienced mathematician
-- in attempting to simplify it may incorrectly believe that 49/98 = 4/8, which
-- is correct, is obtained by cancelling the 9s.
--
-- We shall consider fractions like, 30/50 = 3/5, to be trivial examples.
--
-- There are exactly four non-trivial examples of this type of fraction,
-- less than one in value, and containing two digits in the numerator
-- and denominator.
--
-- If the product of these four fractions is given in its lowest common
-- terms, find the value of the denominator.

isCurious :: (Int, Int) -> Bool
isCurious (numerator, denominator)
    | n1 == d2 && d1 /= 0 && (n2 / d1) == fraction = True
    | n2 == d1 && d2 /= 0 && (n1 / d2) == fraction = True
    | n1 == d1 && d2 /= 0 && (n2 / d2) == fraction = True
    | otherwise = False
    where
        n1 = realToFrac $ numerator `div` 10
        n2 = realToFrac $ numerator `mod` 10
        d1 = realToFrac $ denominator `div` 10
        d2 = realToFrac $ denominator `mod` 10
        fraction = (realToFrac numerator) / (realToFrac denominator)

-- multiply fractions
mul :: (Int, Int) -> (Int, Int) -> (Int, Int)
mul (n1,d1) (n2, d2) = (n1*n2, d1*d2)

-- simplify fraction
simplify :: (Int, Int) -> (Int, Int)
simplify (numerator, denominator) = (numerator `div` g, denominator `div` g)
    where
        g = gcd numerator denominator

main = print . snd . simplify . foldl1 mul $ curious
    where
        curious = filter isCurious $ [(i,j) | j <- [10..99], i <- [10..j-1]]
